src/transform/src/dataflow.rs

```diff
@@ -375,3 +377,58 @@ pub fn optimize_dataflow_monotonic(dataflow: &mut DataflowDesc) -> Result<(), Tr
 
     Ok(())
 }
+
+fn optimize_dataflow_index_imports(
+    dataflow: &mut DataflowDesc,
+    indexes: &dyn IndexOracle,
+) -> Result<(), TransformError> {
+    let mut indexes_by_view = HashMap::new();
+    for sink_desc in dataflow.sink_exports.iter() {
+        indexes_by_view
+            .entry(sink_desc.1.from)
+            .or_insert_with(HashSet::new);
+    }
+    for (_, (index_desc, _)) in dataflow.index_exports.iter() {
+        indexes_by_view
+            .entry(index_desc.on_id)
+            .or_insert_with(HashSet::new);
+    }
+    for build_desc in dataflow.objects_to_build.iter_mut() {
+        build_desc.plan.as_inner_mut().visit_post(&mut |e| {
+            if let MirRelationExpr::Get {
+                id: mz_expr::Id::Global(id),
+                ..
+            } = e
+            {
+                indexes_by_view.entry(*id).or_insert_with(HashSet::new);
+            } else if let MirRelationExpr::ArrangeBy { input, keys } = e {
+                if let MirRelationExpr::Get {
+                    id: mz_expr::Id::Global(id),
+                    ..
+                } = &**input
+                {
+                    for key in keys {
+                        if indexes.indexes_on(*id).find(|k| k == &key).is_some() {
+                            indexes_by_view.get_mut(id).unwrap().insert(key.clone());
+                        }
+                    }
+                }
+            }
+        })?;
+    }
+    for (id, keys) in indexes_by_view.iter_mut() {
+        if keys.is_empty() {
+            if let Some(key) = indexes.indexes_on(*id).next() {
```

Comment:
Maybe if there was no `ArrangeBy` on it, then we are creating it ourselves, so there is only one index? If yes, could you please add a comment and maybe an assert that we have only one index?

Reply:
No, the one index is it's not because we are creating an index ourselves. It's that we only need one index to get all the rows of the view, and if we have not picked any index to query the rows from, we pick one arbitrarily.

I have added a comment that says
```
            // If a dataflow uses a view that has indexes, but it has not
            // chosen to use any specific index of that view, pick an arbitrary
            // index to read from.
```

Reply:
Ah, ok, sorry, I missed the added comment!